### PR TITLE
feat(gatsby): Move webpackCompilationHash into app-data.json

### LINKED ISSUE #15872: Can you opt-out of webpackCompilationHash (to improve Netlify build times)
## Summary
Within the `gatsby-script-loader` script at the bottom of any page there is a `window.webpackCompilationHash` set. This hash is updated during every change to anything in the `src` directory.

This means deploying onto services such as Netlify take longer than expected. Netlify has advanced features for builds so only assets which content has updated since the last build are transferred to the underlying CDN. However, because the `webpackCompilationHash` is updated if anything changes in `src` this means all files will be transferred for only a small change.

For example:

1. `gatsby build` - To give an initial built site.
2. Change the `404` template.
3. `gatsby build` - Builds site but all outputted pages are changed because a single file has been updated.

For larger sites generating 1000s of pages this is a real problem. An update to a fairly static page will result in the dynamic pages also being updated.

Is it possible that we could add an option for the hash to not be generated an inlined in HTML output?

How integral is it to the inner workings of Gatsby?

## Relevant information

Mainly affects large sites being deployed on services such as Netlify.

### PR DESCRIPTION
**Pull request by afenton90**
## Description

`webpackCompilationHash` is currently used in `page-data.json` & HTML files to decipher whether the Gatsby app running in the browser is at the same version as the code on the server/CDN.

At the moment this hash changes if anything within `./src` has changed between builds. Meaning a 404 file could be changed which would result in all `page-data.json` & HTML files being updated even if they have not changed between builds.

This causes some issues when using some cloud CDN providers as described in this issue #15872 .

This PR now moves the `webpackCompilationHash` to it's own `app-data.json` which is loaded at the same time as page-data in order to reduce the blast radius of changes in the build output.
This allows for diffs to take place between what is hosted in a CDN and what is about to be pushed to a CDN.

## Related Issues

close #15872

### CODE REVIEW

**Comment by afenton90** [comment]
`starters_validate` check seems to be failing because it cannot access the audit npm endpoint. Status page shows no downtime though.

**Comment by pieh** [comment]
> `starters_validate` check seems to be failing because it cannot access the audit npm endpoint. Status page shows no downtime though.

Yeah, those tests seems to be flaky because of audit randomly not working sometimes. I restarted that one test, just to (hopefully) not report false negative here.

**Comment by sidharthachatterjee** [comment]
TODO

- [ ] @sidharthachatterjee to publish this under a tag 
- [ ] Add a section to https://www.gatsbyjs.org/docs/caching/
- [ ] @afenton90 to write a blog post about approach here and benefits (separate and not a blocker)

**Comment by sidharthachatterjee** [comment]
Published under `gatsby@app-data`

**Comment by KyleAMathews** [comment]
Can you publish two versions of a moderately complicated site with and without this PR somewhere and do a visual comparison test of the sites on https://www.webpagetest.org/video/?

The main worry I have about this approach is it slows down the initial page load a bit as it's yet another file we're adding to the initial load. This could slow things own as browsers place limits on how many sites you can load in parallel. It'd be good to have hard numbers to see. Maybe also run some tests with other browsers (the visual comparison only has chrome).

If this does prove a problem, we could maybe lazy load the file later as it's not needed initially right?  Or maybe just do this right away and ignore the test haha 🤔

**Comment by afenton90** [comment]
Hey @KyleAMathews I've published a version of the `@vojtaholik/gatsby-theme-simplecast` with and without the change in this PR and run the visual regression test in WPT.
Results are here https://www.webpagetest.org/video/compare.php?tests=190821_EF_7d163ab01ee4e66eb0975ae06a54f1de,190821_05_42324dc66644b02ee5d8e684f2b8fc76

You can see that even though one more request is happening there is no affect on performance of the site. This is mainly down to the small size of `app-data.json` but also the pre-fetching that is done for it.

The sites tested are:
**Before App Data:** https://determined-pike-5c14a7.netlify.com/
**After App Data:** https://5d5d1badc5c9700008a7a472--determined-pike-5c14a7.netlify.com/

**Comment by afenton90** [comment]
> Add delete for app-data in packages/gatsby/src/bootstrap/index.js

@siddharthkp after looking at this point we should not need it as the `app-data.json` file is overridden on every build.

**Comment by sidharthachatterjee** [comment]
@afenton90 Yup, makes sense. Also, wrong Siddharth 😅

**Comment by afenton90** [comment]
Haha sorry @sidharthachatterjee 😆

**Comment by me4502** [comment]
On a site with a lot of images and other files that need to be downloaded - it appears to cause a 0.3 second delay.

https://www.webpagetest.org/video/compare.php?tests=190822_PH_e38944357bffb30958a92aff2cb3d95a%2C190822_6A_bf4c128459c20694edb1b51a235b6aec&thumbSize=200&ival=100&end=visual

This is probably a pretty extreme example, and in our case the benefits outweigh the slight slowdown.

**Comment by KyleAMathews** [comment]
0.3 seconds is pretty big difference.

To repeat my earlier question — can we lazy load the app-data.json file so there's no perf implications?

**Comment by afenton90** [comment]
Looking at the trace for that test it seems the difference in speed is mostly down to initial load of html and not other resources.

I think given the size of the app-data file and that it’s loaded alongside the page-data file we shouldn’t see degraded performance of sites.

**Comment by KyleAMathews** [comment]
Browsers restrict how many parallel requests they can make so this will cause trouble for some sites.

Can we lazy load the file? I haven't looked at the implementation mich so don't know if it's possible but this would be my preference.

File: packages/gatsby/cache-dir/static-entry.js [hunk 1]
```diff
@@ -322,6 +322,15 @@ export default (pagePath, callback) => {
       )
     })
 
+  headComponents.push(
+    <link
```
Comment (sidharthachatterjee) [inline packages/gatsby/cache-dir/static-entry.js]: Is the preload here causing the slow down?

**Comment by dtbaker** [comment]
This is exciting @afenton90 ! I'm going to try this on our `mixkit` website. This should let us keep deploy times low as we grow to thousands of pages.

**Comment by afenton90** [comment]
Hey @sidharthachatterjee & @KyleAMathews tried to reach out to you separately but thought it may be best to leave a message here.
Happy to look at lazy loading the `app-data.json` file.
I can remove the `preload` at the top, then simply let it download at the same time as the `page-data.json` file is requested, this should have the desired effect of lazy loading as the script is not executed until the bottom of page `body`.

Does this sound reasonable? I'd love to get this PR in soon.

**Comment by afenton90** [comment]
Oh and no worries @dtbaker fat fingers forgiven 😜 (sounds like some weird Harry Potter spell, but you get me)

**Comment by KyleAMathews** [comment]
For lazy loading, we would load it after the page is loaded. Perhaps just have the page prefetching logic load it?

**Comment by afenton90** [comment]
Cool. So the page prefetching logic is already covered and the `app-data.json` file will be loaded when `loadPage` is called for a pre-fetch.

For the initial entry to a Gatsby site is where I want to be clear on where to load the file.

So should lazy loading be on a page load event of `body` loaded?
Or because the JS files are the last to be loaded anyway, just remove the `preload` and fetch the file initially when `page-data.json` is processed?

**Comment by KyleAMathews** [comment]
We want to wait until the normal js files are loaded so this file doesn't delay any of them loading.

Since page prefetching already loads this file, let's just wait until page prefetching starts.

**Comment by afenton90** [comment]
The `preload` for `app-data.json` has now been removed and the file is now lazy loaded.

**Comment by afenton90** [comment]
Would be good to get a final (hopefully) review on this once one of the core team has chance.

**Comment by ArthurHwang** [comment]
I am so anxiously awaiting this patch.  Thanks to all of you for working on this.  My project deploy times to netlify are seriously an issue, I am glad to see there is hope in sight.

**Comment by tmilewski** [comment]
Hey, what's the best way to help push this forward? This change would be a massive benefit for us.

At this point, I've been trying to test the PR out, but `yarn` seems to be choking all over the place when trying to install it. Additionally, the branch is >200 commits behind `master`.

@afenton90 @sidharthachatterjee @KyleAMathews

**Comment by ArthurHwang** [comment]
@tmilewski I am also anxiously awaiting this.  I check this thread every day for any updates.  My project workflow right now is really hampered due to this.

**Comment by afenton90** [comment]
Hey @tmilewski & @ArthurHwang, the only thing remaining on the PR is to manually test it against some larger projects.
I know that @sidharthachatterjee was planning to do this against gatsbyjs.org & reactjs.org. However, I'm also aware he's had a lot going on at the moment, so probably hasn't got around to this yet.

If either of you have some larger projects you could test against that would be great! I know yarn gets funny about branch names sometimes, so the best way may be to clone the repo locally, build it, and do a relative path install in your project to test.

I know that one of the core maintainers should test against a larger site also. @sidharthachatterjee @KyleAMathews would either of you be able to arrange this?

**Comment by afenton90** [comment]
@sidharthachatterjee any idea when this can be tested against the 2 sites that we have mentioned above?

**Comment by ArthurHwang** [comment]
> @sidharthachatterjee any idea when this can be tested against the 2 sites that we have mentioned above?

@afenton90 I've never actually tested a pull request / branch of a package so I am not really sure how to do it especially in this case to test this specific version of gatsby created by you, but if you can give me a quick run through how to test this I would greaty appreciate it.

I have a work site I rebuilt with gatsby that is pretty large, I think.  I would love to help out with this.
It has around 6100 pages, 4000 images.

**Comment by afenton90** [comment]
@ArthurHwang if you follow these steps you should be able to bring down this branch and run against your site locally.

1. Follow the steps in this guide https://www.gatsbyjs.org/contributing/setting-up-your-local-dev-environment/ & confirm you can do a build.
1. Clone this branch and do another build
1. Link to your local app using `gatsby-dev`
1. Build & serve the site. You will be able to validate you are running on this branch by seeing an `app-data.json` file being downloaded in the network tab.

Let me know how you get on & if you have any sticking points getting this to run 👍

**Comment by ArthurHwang** [comment]
> @ArthurHwang if you follow these steps you should be able to bring down this branch and run against your site locally.
> 
> 1. Follow the steps in this guide https://www.gatsbyjs.org/contributing/setting-up-your-local-dev-environment/ & confirm you can do a build.
> 2. Clone this branch and do another build
> 3. Link to your local app using `gatsby-dev`
> 4. Build & serve the site. You will be able to validate you are running on this branch by seeing an `app-data.json` file being downloaded in the network tab.
> 
> Let me know how you get on & if you have any sticking points getting this to run 👍

Thank you @afenton90 , will get on it in the next day or two.

**Comment by afenton90** [comment]
There seems to be a gap now between the nav bar and the menu. Would you be able to take a look?
Screenshot 2019-10-16 at 15 23 25

**Comment by afenton90** [comment]
React.org test site deployed with this version of Gatsby.
https://boring-shockley-d8cae0.netlify.com/

You should be able to see the `app-data.json` files in the Network Tab.

**Comment by tmilewski** [comment]
I'm running into this specifically with `gatsby@app-data`:

```
› rm -rf .cache; rm -rf public; yarn develop
yarn run v1.17.3
$ gatsby develop
The above error occurred in the <Provider> component:
    in Provider (created by ReduxStoreProvider)
    in ReduxStoreProvider
    in App

React will try to recreate this component tree from scratch using the error boundary you provided, App.
Warning: App: Error boundaries should implement getDerivedStateFromError(). In that method, return a state update to display an error message or fallback UI.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by afenton90** [comment]
@tmilewski thanks for taking the time to try this PR out.
I've just tried doing a fresh clone of my reactjs.org clone and running `develop` & `build` but afraid I don't see this error.

Nothing in this PR has touched any `Provider` code, but I'm aware this doesn't resolve your issue.
Happy to jump on a call though to ensure there is no issue specifically from this PR.

**Comment by sidharthachatterjee** [comment]
@tmilewski This might be related to some recent changes in gatsby-cli (which uses Ink and therefore React) 

Can you open an issue for that with a reproduction please?

File: packages/gatsby/src/utils/app-data.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+const fs = require(`fs-extra`)
+const path = require(`path`)
+
+const APP_DATA_JSON = `app-data.json`
+
+const write = (publicDir, hash) => {
+  fs.writeJson(path.join(publicDir, APP_DATA_JSON), {
+    appHash: hash,
```
[the remark this answers is not in the document]
Comment (afenton90) [reply packages/gatsby/src/utils/app-data.js]: Happy to change back, not particularly attached to the name.

File: packages/gatsby/cache-dir/loader.js [hunk 3]
```diff
@@ -305,6 +312,30 @@ export class BaseLoader {
     const page = this.pageDb.get(pagePath)
     return page && page.notFound === true
   }
+
+  loadAppData() {
```
[the remark this answers is not in the document]
Comment (afenton90) [reply packages/gatsby/cache-dir/loader.js]: Yeah can add in a retry. Will do the same retry 3 times.

File: packages/gatsby/src/utils/app-data.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+const fs = require(`fs-extra`)
+const path = require(`path`)
+
+const APP_DATA_JSON = `app-data.json`
```
[the remark this answers is not in the document]
Comment (afenton90) [reply packages/gatsby/src/utils/app-data.js]: Yeah that's a great point actually. I'll move it to be at the root of the `page-data` directory.

[the remark this answers is not in the document]
Comment (sidharthachatterjee) [reply packages/gatsby/src/utils/app-data.js]: Okay, thank you!

**Comment by afenton90** [comment]
@sidharthachatterjee I've made all the requested changes.
The page nav as an integration test under `long-term-caching` which tests the reload behaviour, so as long as that passes we're all good.

**Comment by sidharthachatterjee** [comment]
@afenton90 Amazing work! I've published the latest in `gatsby@app-data`

Some snap shots failing in https://app.circleci.com/jobs/github/gatsbyjs/gatsby/255360

**Comment by tmilewski** [comment]
@sidharthachatterjee @afenton90 Thanks for the help, and yes, it appears to be `gatsby-cli`. If we could update this branch to match the latest release, I *think* that this should resolve itself. I'll try and carve out some time for a reproduction, but I can't guarantee much at this very moment.

In the meantime, here's what I do have:

---

For reference, this PR is using the following:
`gatsby@2.13.66`
`gatsby-cli@2.8.1`

```
> yarn add gatsby@app-data
    # gatsby-cli is 2.8.1
    # FAILURE

> yarn add gatsby@2.13.66
    # 2.13.66 maps to the version used in PR
    # gatsby-cli  is ^2.7.30
    # SUCCESS

> yarn add gatsby@2.13.66 gatsby-cli@2.8.1
> yarn serve  
    # FAILURE

> yarn add gatsby@2.13.66 gatsby-cli@2.8.0
› yarn serve  
    # FAILURE

> yarn add gatsby@2.13.66 gatsby-cli@2.7.59
    # SUCCESS
```

**Comment by sidharthachatterjee** [comment]
Thank you so much for all your work on this, @afenton90 💜 

Published in `gatsby@2.17.0`

**Comment by rroslaniec** [comment]
Would be nice to update [gatsby-docker](https://github.com/gatsbyjs/gatsby-docker) to add cache-control for app-data.json file.

**Comment by Skillz4Killz** [comment]
Wanted to just say thank you for the work on this. This has reduced the changed files by half for me. Although im still curious why it still thinks there are 300+ files modified when deploying to netlify. @afenton90 I am curious if you think there is anything more that can be added to this  to stop gatsby recreating new files. Both images below come from just simple dependabot PR merges.

Before this merge:
image

After the merge:
image

**Comment by afenton90** [comment]
> Would be nice to update gatsby-docker to add cache-control for app-data.json file.
@rroslaniec yeah that's a good point.

@Skillz4Killz yeah I think there is even more that be done when it comes to deployments on Netlify. For example, the contenthash on assets is not needed when deploying the site to Netlify. Netlify will handle the cache invalidation for assets in different ways. I plan to put in a PR for `gatsby-plugin-netlify` towards the end of this week to resolve this.

**Pull request merged**